### PR TITLE
Recipient reputation

### PR DESCRIPTION
**Pull request by shahafn**
Adding default (pretty naive) recipient reputation mechanism.
This PR is based on branch review-fixes.  DO NOT MERGE UNTIL #117 IS MERGED

### CODE REVIEW

File: server/src/librelay/relay_server.go [hunk 1]
```diff
@@ -414,6 +401,16 @@ func (relay *RelayServer) CreateRelayTransaction(request RelayTransactionRequest
 		return
 	}
 
+	// Check that the recipient reputation is not too low to serve it
+	reputationScore, err := relay.ReputationStore.GetScore(&request.To)
+	if err != nil {
+		log.Println("Error in retrieving recipient's reputation from database:", err)
+		log.Println("Ignoring recipient's reputation...")
+	} else if reputationScore.Cmp(big.NewInt(ReputationThreshold)) <= 0 {
+		log.Printf("Reputation of %s too low: %d", request.To.Hex(), reputationScore.Int64())
```
[the remark this answers is not in the document]
Comment (drortirosh) [reply server/src/librelay/relay_server.go]: **more important**: once the client's reputation is too low, that relay will stop serving it - forever (that is, until the relay is restarted, and the repuration DB is cleared)

[the remark this answers is not in the document]
Comment (drortirosh) [reply server/src/librelay/relay_server.go]: Another possible bug: reputationScore<=0 means "don't serve" - but that's exactly the score of a new contract. how can it serve anything?

File: contracts/RelayHub.sol [hunk 2]
```diff
@@ -199,36 +201,35 @@ contract RelayHub is IRelayHub {
...
         uint charge = (gasOverhead + initialGas - gasleft()) * gasPrice * (100 + transactionFee) / 100;
-        if (!successPost) {
-            emitTransactionRelayed(msg.sender, from, to, encodedFunction, uint(RelayCallStatus.PostRelayedFailed), charge);
-        } else {
-            emitTransactionRelayed(msg.sender, from, to, encodedFunction, uint(status), charge);
-        }
-        require(balances[to] >= charge, "insufficient funds");
+        emitTransactionRelayed(msg.sender, from, to, encodedFunction, uint(status), charge);
+        // We already checked at the beginning that the recipient has enough balance. This is more
+        require(balances[to] >= charge, "Should not get here");
```
[the remark this answers is not in the document]
Comment (shahafn) [reply contracts/RelayHub.sol]: Fixed in #122  ( added a comment explaining this require)

File: contracts/RelayHub.sol [hunk 3]
```diff
@@ -199,36 +201,35 @@ contract RelayHub is IRelayHub {
...
         // Relay transactionFee is in %.  E.g. if transactionFee=40, payment will be 1.4*usedGas.
         uint charge = (gasOverhead + initialGas - gasleft()) * gasPrice * (100 + transactionFee) / 100;
-        if (!successPost) {
-            emitTransactionRelayed(msg.sender, from, to, encodedFunction, uint(RelayCallStatus.PostRelayedFailed), charge);
-        } else {
-            emitTransactionRelayed(msg.sender, from, to, encodedFunction, uint(status), charge);
-        }
-        require(balances[to] >= charge, "insufficient funds");
+        emitTransactionRelayed(msg.sender, from, to, encodedFunction, uint(status), charge);
+        // We already checked at the beginning that the recipient has enough balance. This is more
```
[the remark this answers is not in the document]
Comment (shahafn) [reply contracts/RelayHub.sol]: Fixed in #122

File: contracts/RelayHub.sol [hunk 4]
```diff
@@ -199,36 +201,35 @@ contract RelayHub is IRelayHub {
      */
     function relayCall(address from, address to, bytes memory encodedFunction, uint transactionFee, uint gasPrice, uint gasLimit, uint nonce, bytes memory approval) public {
         uint initialGas = gasleft();
+        require(balances[to] >= gasPrice * initialGas, "Recipient balance too low");
```
[the remark this answers is not in the document]
Comment (shahafn) [reply contracts/RelayHub.sol]: In relay_server.go I detail what the total gasLimit for this transaction

File: server/src/librelay/relay_server.go [hunk 5]
```diff
@@ -483,21 +494,75 @@ func (relay *RelayServer) CreateRelayTransaction(request RelayTransactionRequest
...
 	}
-	log.Println("Estimated max charge of relayed tx:",maxCharge)
+	// Maximum gasLimit of relayed tx consists of:
+	// 1. request.GasLimit - user request gasLimit for the relayed function call
+	// 2. gasOverhead - Gas cost of all relayCall() instructions before first gasleft() and last gasleft()
+	// 3. gasReserve - Gas cost of all relayCall() instructions after first gasleft() and before last gasleft()
+	// 4. acceptRelayedCallMaxGas & postRelayedCallMaxGas - max gas cost of recipient calls acceptRelayedCall() & postRelayedCall()
+	gasLimit := big.NewInt(0)
+	gasLimit.Add(&request.GasLimit, gasOverhead)
+	gasLimit.Add(gasLimit, gasReserve)
```
[the remark this answers is not in the document]
Comment (shahafn) [reply server/src/librelay/relay_server.go]: The client's gasLimit is `request.GasLimit` so it's not missing

File: server/src/librelay/relay_server.go [hunk 6]
```diff
@@ -483,21 +494,75 @@ func (relay *RelayServer) CreateRelayTransaction(request RelayTransactionRequest
...
+	gasLimit.Add(gasLimit, acceptRelayedCallMaxGas)
+	gasLimit.Add(gasLimit, postRelayedCallMaxGas)
+	log.Println("Estimated max charge of relayed tx:", maxCharge, "GasLimit of relayed tx:",gasLimit)
 
 	signedTx, err = relay.sendDataTransaction(
 		fmt.Sprintf("Relay(from=%s, to=%s)", request.From.Hex(), request.To.Hex()),
 		func(auth *bind.TransactOpts) (*types.Transaction, error) {
-			auth.GasLimit = gasLimit.Add(&request.GasLimit, gasReserve).Add(gasLimit, gasReserve).Uint64()
+			auth.GasLimit = gasLimit.Uint64()
```
[the remark this answers is not in the document]
Comment (shahafn) [reply server/src/librelay/relay_server.go]: Types of all integers returned from contracts are `big.Int`, so I keep it that way. FWIW, I think it should be fixed in auth implementation i.e it should be `big.Int`

File: server/src/librelay/relay_server.go [hunk 7]
```diff
@@ -483,21 +494,75 @@ func (relay *RelayServer) CreateRelayTransaction(request RelayTransactionRequest
...
+		log.Println(err)
+		return
+	}
+	parsed, err := abi.JSON(strings.NewReader(librelay.RelayHubABI))
+	if err != nil {
+		log.Println(err)
+		return
+	}
+	transactionRelayedEvent := new(librelay.RelayHubTransactionRelayed)
+	boundhub := bind.NewBoundContract(relay.HubAddress(), parsed, nil, nil, nil)
```
[the remark this answers is not in the document]
Comment (shahafn) [reply server/src/librelay/relay_server.go]: Right, Fixed

**Pull request closed without merging**